sklearn/impute/_base.py

```diff
@@ -555,6 +557,11 @@ def transform(self, X):
         check_is_fitted(self)
 
         X = self._validate_input(X, in_fit=False)
+        # we use min type for the fill_value to avoid changing the input dtype
+        # if not necessary
+        new_dtype = np.promote_types(X.dtype, self._fill_value_min_type)
+        if new_dtype != X.dtype:
+            X = X.astype(new_dtype, copy=self.copy)
```

Comment:
Good point. I don't see a good way to do that, the api of check_array is too limited for this situation. Let's keep the potential double copy for now in the hope of a better check_array some day. It should not happen very often.

However, let's not use self.copy here since it has a specific meaning. Instead we should let numpy assk numpy to not make a copy if possible (it will make one if it has to anyway)

```suggestion
        new_dtype = np.promote_types(X.dtype, self._fill_value_min_type)
        X = X.astype(new_dtype, copy=False)
```